### PR TITLE
Improve `BoundedVec` API

### PR DESCRIPTION
**Pull request by shawntabrizi**
This adds some useful APIs for making the conversion from `Vec` to `BoundedVec` more seamless.

### CODE REVIEW

File: frame/support/src/storage/bounded_vec.rs [hunk 1]
```diff
@@ -188,11 +201,26 @@ impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Deref for BoundedVec<T, S> {
 	}
 }
 
+impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::DerefMut for BoundedVec<T, S> {
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_vec.rs:204]: This allows you to call methods like `extend` and `push` without checking, not a good idea. 

overall, bounded vec should only coerce to vec when methods are **safe regarding the bound** (cannot expand and re allocate). This is why I only did deref initially. 

Same comments apply to indexMut and AsRefMut: if you can expand the size with them, we can't add them and instead it should be a checked method on the type itself, or just manually convert to vec and try and convert it back.

**Review by kianenigma — CHANGES_REQUESTED** [review]
I think adding derefMut can be dangerous and add footguns.

File: frame/support/src/storage/bounded_vec.rs (hunk 1 above)
Comment (bkchr) [reply frame/support/src/storage/bounded_vec.rs:204]: I'm going with DerefMut, this shouldn't be implelemented as said by @kianenigma

However, IndexMut and AsMut are safe as they don't support extending the vec. Both just give you multiple access to the vec elements.

File: frame/support/src/storage/bounded_vec.rs [hunk 2]
```diff
@@ -188,11 +201,26 @@ impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Deref for BoundedVec<T, S> {
...
+    fn deref_mut(&mut self) -> &mut Vec<T> {
+        &mut self.0
+    }
+}
+
```
Comment (thiolliere) [inline frame/support/src/storage/bounded_vec.rs:209]: I agree let's remove deref_mut
```suggestion
```

**Review by thiolliere — COMMENTED** [review]
once we remove deref mut, then it is good to me

**Review by kianenigma — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

File: frame/support/src/storage/bounded_vec.rs [hunk 3]
```diff
@@ -189,10 +202,19 @@ impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Deref for BoundedVec<T, S> {
...
+	#[inline]
+	fn index(&self, index: I) -> &Self::Output {
+		Index::index(&**self, index)
+	}
+}
+
+impl<T: BoundedVecValue, S: Get<u32>, I: SliceIndex<[T]>> IndexMut<I> for BoundedVec<T, S> {
+	#[inline]
+	fn index_mut(&mut self, index: I) -> &mut Self::Output {
+		IndexMut::index_mut(&mut **self, index)
```
Comment (bkchr) [inline frame/support/src/storage/bounded_vec.rs:217]:
```suggestion
		self.0.index_mut(index)
```

File: frame/support/src/storage/bounded_vec.rs [hunk 4]
```diff
@@ -189,10 +202,19 @@ impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Deref for BoundedVec<T, S> {
...
-impl<T: BoundedVecValue, S: Get<u32>> sp_std::ops::Index<usize> for BoundedVec<T, S> {
-	type Output = T;
-	fn index(&self, index: usize) -> &Self::Output {
-		self.get(index).expect("index out of bound")
+impl<T: BoundedVecValue, S: Get<u32>, I: SliceIndex<[T]>> Index<I> for BoundedVec<T, S> {
+	type Output = I::Output;
+
+	#[inline]
+	fn index(&self, index: I) -> &Self::Output {
+		Index::index(&**self, index)
```
Comment (bkchr) [inline frame/support/src/storage/bounded_vec.rs:210]:
```suggestion
		self.0.index(index)
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**